### ISSUE TITLE
PDN short circuiting Metal1 in endcap/tbtie cells

### ISSUE THREAD

**Issue by gatecat** [The-OpenROAD-Project/OpenROAD#4857]
### Description

We are using OpenROAD (with ORFS) to tape out on a proprietary PDK. The cell library contains endcap, corner and tbtie cells that contain a metal1 ring (that needs to be connected to VSS in our case).

Those cells are placed fine using the `tapcell` command, but the metal1 PDN stripes generated by `add_pdn_stripe -followpins` don't take this into account and short onto the metal1.

Is there a way of avoiding skipping the first/last metal1 stripe and trimming the length of the stripes, such that they don't short the metal1 inside this tap cell ring, other than ad hoc patching the PDN generation?

Incidentally, what would be the nicest way to connect a pin in every cell in this ring to VSS, and generate the necessary routing for that?

**Comment by maliberty** [comment]
A few pictures would help as I don't understand your description.  Normally tapcells/endcaps have power pins in the same location as any other standard cell and follow pins isn't an issue.

**Comment by gatecat** [comment]
Please excuse my "not to scale" sketch to avoid violating any NDAs, hope this still helps.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/78621419/99e51c40-7cc1-4291-a640-e748e0bfb78e)

As well as the standard cell power strips, there is also a "guard ring" that connects a continuous ring of metal1 (pins called DTI in the LEF) around the edge of the core area - this needs to be connected to P-Substate bias e.g. Vss - but at the moment, it's also shorted to the extended Vdd pins.

**Comment by maliberty** [comment]
That's quite odd, if I look at the gf180 endcap I see:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/14cedac2-8616-4f64-9262-d0081f52910a)

The rails are not going to be shorted due to the gap.  Are you saying your endcap cells has a metal strap between the rails?

**Comment by maliberty** [comment]
In the gds for m1:
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/def4209f-a248-4feb-8989-772805970fd4)

**Comment by gatecat** [comment]
Nope - the rails inside the cell end well before that Metal1 ring (which is outside the placement boundary of the cell) - it's OpenROAD that's extending the rails such that they short the outer ring.

**Comment by maliberty** [comment]
How did you place cells outside the core area?  tapcell will not normally do that.

**Comment by maliberty** [comment]
The core area is defined by the bbox of all the rows in the design.  I'm not sure what "placement boundary" is exactly.

**Comment by maliberty** [comment]
@gatecat is there anything he we can resolve or should we just close?

**Comment by gatecat** [comment]
Yes - as far as cell content ending up outside of the core area, this happens because of negative coordinates in the LEF, even though the placement origin is within the core area. You should be able to see the TBCAPTIE/ENDCAPTIE/CNRCAPTIE cells in the LEF/GDS in our private repo, let me know if you want any other info either here or privately.

**Comment by maliberty** [comment]
stale
